File: 클래스수업/1.rb
class Cal
  attr_reader :v1, :v2
  attr_writer :v1
  @@_history = []
  def initialize(v1,v2)
    @v1 = v1
    @v2 = v2
  end
  def add()
    result = @v1+@v2
    @@_history.push("add : #{@v1}+#{@v2}=#{result}")
    return result
  end
  def subtract()
    result = @v1-@v2
    @@_history.push("subtrat : #{@v1}-#{@v2}=#{result}")
    return result
  end
  def Cal.history()
    for item in @@_history
      p item
    end
  end
end
class CalMultiply < Cal
  def multiply()
    result = @v1-@v2
    @@_history.push("multiply : #{@v1}*#{@v2}=#{result}")
    return result
  end
end
class CalDivice < CalMultiply
  def devide()
    result = @v1/@v2
    @@_history.push("devide : #{@v1}/#{@v2}=#{result}")
    return result
  end
end


c1 = CalMultiply.new(10,10)
p c1.add()
p c1.subtract()
p c1.multiply()
c2 = CalMultiply.new(30,20)
p c2.add()
p c2.multiply()
c3 = CalDivice.new(20,10)
p c3.devide()

# p c2.add()
Cal.history()

File: Class/1.py
class Cal(object):
    _history = []
    def __init__(self,v1,v2):
        self.v1 = v1
        self.v2 = v2
    def add(self):
        result = self.v1+self.v2
        Cal._history.append("add:%d+%d=%d" % (self.v1, self.v2, result))
        return result
    def subtract(self):
        result = self.v1-self.v2
        Cal._history.append("subtract:%d-%d=%d" % (self.v1,self.v2,result))
        return result
    @classmethod
    def history(cls):
        for item in Cal._history:
            print(item)
class CalMultiply(Cal):
    def mutiply(self):
         result = self.v1*self.v2
         Cal._history.append("multiply:%d*%d=%d" % (self.v1,self.v2,result))
         return result
class CalDivice(CalMultiply):
    def devide(self):
        result = self.v1/self.v2
        Cal._history.append("devide:%d/%d=%d" % (self.v1,self.v2,result))
        return result

c1 = Cal(10,10)
print(c1.add())
print(c1.subtract())
# c1.v2 = 30
# c1.setV1('one')
print(c1.add())
print(c1.subtract())
Cal.history()

# c2 = Cal(30,20)
# print(c2.add())
# print(c2.subtract())
